tests/integration/test_config.py

```diff
@@ -339,3 +378,381 @@ async def test_host_group_config(cluster: Cluster) -> None:
+
+        # get two copies of the same object with different requesters
+        service_name = "complex_config"
+        service_1 = await cluster.services.get(name__eq=service_name)
+        service_2 = await (await second_client.clusters.get(name__eq=cluster.name)).services.get(name__eq=service_name)
+
+        component = await service_1.components.get(name__eq="component1")
+        mapping = await cluster.mapping
+        await mapping.add(component=component, host=host)
+        await mapping.save()
+
+        # set required field
+        cfg = await service_1.config
+        cfg["Set me", Parameter].set(True)
+        await cfg.save()
+
+        # tests
+        await two_sessions_case_1(service_1, service_2, httpx_client=httpx_client)
+        await two_sessions_case_2(service_1, service_2, httpx_client=httpx_client)
+        await two_sessions_case_3(service_1, service_2, httpx_client=httpx_client)
+        await two_sessions_case_4(service_1, service_2, httpx_client=httpx_client)
+        await two_sessions_case_5(service_1, service_2, httpx_client=httpx_client)
+        await two_sessions_case_6(service_1, service_2, host=host)
+        await two_sessions_case_7(service_1, service_2, host=host)
+        await two_sessions_case_8(service_1, service_2, httpx_client=httpx_client)
+        await two_sessions_case_9(service_1, service_2, httpx_client=httpx_client)
+
+
+async def two_sessions_case_1(obj1: Service, obj2: Service, httpx_client: AsyncClient) -> None:
+    """
+    user 1: set value (same as initial), save config;
+    user 2: set value (same as initial), refresh (apply remote)
+    """
+
+    cfg1, cfg2 = await refresh_and_get_two_configs(obj1, obj2)
+    assert isinstance(cfg1, ObjectConfig) and isinstance(cfg2, ObjectConfig)
+
+    field_display_name, field_name, value = "Complexity Level", "complexity_level", 4
+    assert (await get_object_config(obj1, httpx_client))[field_name] == value
+
+    field1 = cfg1[field_display_name, Parameter]
+    field2 = cfg2[field_display_name, Parameter]
+
+    # user 1
+    field1.set(value)
+    await cfg1.save()
+
+    # user 2
+    field2.set(value)
+    await cfg2.refresh(strategy=apply_remote_changes)
+
+    assert (
+        (await get_object_config(obj1, httpx_client))[field_name]
+        == value
+        == cfg1[field_display_name, Parameter].value
+        == cfg2[field_display_name, Parameter].value
+    )
+
+
+async def two_sessions_case_2(obj1: Service, obj2: Service, httpx_client: AsyncClient) -> None:
+    """
+    user 1: activate group, set field1's value, save config;
+    user 2: activate group, set field2's value, refresh (apply local); save config
+    """
+
+    cfg1, cfg2 = await refresh_and_get_two_configs(obj1, obj2)
+    assert isinstance(cfg1, ObjectConfig) and isinstance(cfg2, ObjectConfig)
+
+    agr_display_name, agr_name, field1, value1, field2, value2 = "Optional", "agroup", "justhere", 3, "field2", 6
```

Comment:
imo difficult to track values

one per line will be nicer

Reply:
refactored